// @payloadcms/plugin-cloud/dist/staticHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "getStaticHandler", {
    enumerable: true,
    get: function() {
        return getStaticHandler;
    }
});
const _createKey = require("./utilities/createKey");
const _getStorageClient = require("./utilities/getStorageClient");
const getStaticHandler = ({ cachingOptions, collection })=>{
    let maxAge = 86400 // 24 hours default
    ;
    let collCacheConfig;
    if (cachingOptions !== false) {
        // Set custom maxAge for all collections
        maxAge = cachingOptions?.maxAge || maxAge;
        collCacheConfig = cachingOptions?.collections?.[collection.slug] || {};
    }
    // Set maxAge using collection-specific override
    maxAge = collCacheConfig?.maxAge || maxAge;
    const cachingEnabled = cachingOptions !== false && !!process.env.PAYLOAD_CLOUD_CACHE_KEY && collCacheConfig?.enabled !== false;
    return async (req, res, next)=>{
        const filename = req.params.filename;
        let fileKeyWithPrefix = '';
        if (!filename) {
            req.payload.logger.warn({
                msg: `No filename provided for static file against collection: ${collection.slug}`
            });
        }
        try {
            const { identityID, storageClient } = await (0, _getStorageClient.getStorageClient)();
            fileKeyWithPrefix = (0, _createKey.createKey)({
                collection: collection.slug,
                filename,
                identityID
            });
            const object = await storageClient.getObject({
                Bucket: process.env.PAYLOAD_CLOUD_BUCKET,
                Key: fileKeyWithPrefix
            });
            res.set({
                'Content-Length': object.ContentLength,
                'Content-Type': object.ContentType,
                ...cachingEnabled && {
                    'Cache-Control': `public, max-age=${maxAge}`
                },
                ETag: object.ETag
            });
            if (object?.Body) {
                return object.Body.pipe(res);
            }
            return next();
        } catch (err) {
            req.payload.logger.error({
                err,
                msg: `Error getting file from cloud storage: '${fileKeyWithPrefix}'`
            });
            return next();
        }
    };
};
