// @payloadcms/db-mongodb/dist/models/buildCollectionSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _mongoosepaginatev2 = /*#__PURE__*/ _interop_require_default(require("mongoose-paginate-v2"));
const _buildQuery = /*#__PURE__*/ _interop_require_default(require("../queries/buildQuery"));
const _buildSchema = /*#__PURE__*/ _interop_require_default(require("./buildSchema"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const buildCollectionSchema = (collection, adapter)=>{
    const schema = (0, _buildSchema.default)(adapter, collection.fields, {
        draftsEnabled: Boolean(typeof collection?.versions === 'object' && collection.versions.drafts),
        indexSortableFields: adapter.payload.config.indexSortableFields,
        options: {
            minimize: false,
            timestamps: collection.timestamps !== false,
            ...adapter.schemaOptions,
            ...adapter.collectionOptions[collection.slug]?.schemaOptions || {}
        }
    });
    if (adapter.payload.config.indexSortableFields && collection.timestamps !== false) {
        schema.index({
            updatedAt: 1
        });
        schema.index({
            createdAt: 1
        });
    }
    schema.plugin(_mongoosepaginatev2.default, {
        useEstimatedCount: true
    }).plugin((0, _buildQuery.default)({
        collectionSlug: collection.slug
    }));
    return schema;
};
const _default = buildCollectionSchema;
