// @payloadcms/db-mongodb/dist/createGlobalVersion.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "createGlobalVersion", {
    enumerable: true,
    get: function() {
        return createGlobalVersion;
    }
});
const _sanitizeInternalFields = /*#__PURE__*/ _interop_require_default(require("./utilities/sanitizeInternalFields"));
const _withSession = require("./withSession");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const createGlobalVersion = async function createGlobalVersion({ autosave, createdAt, globalSlug, parent, req = {}, updatedAt, versionData }) {
    const VersionModel = this.versions[globalSlug];
    const options = await (0, _withSession.withSession)(this, req);
    const [doc] = await VersionModel.create([
        {
            autosave,
            createdAt,
            latest: true,
            parent,
            updatedAt,
            version: versionData
        }
    ], options, req);
    await VersionModel.updateMany({
        $and: [
            {
                _id: {
                    $ne: doc._id
                }
            },
            {
                parent: {
                    $eq: parent
                }
            },
            {
                latest: {
                    $eq: true
                }
            }
        ]
    }, {
        $unset: {
            latest: 1
        }
    }, options);
    const result = this.jsonParse ? JSON.parse(JSON.stringify(doc)) : doc.toObject();
    const verificationToken = doc._verificationToken;
    if (verificationToken) {
        result._verificationToken = verificationToken;
    }
    return (0, _sanitizeInternalFields.default)(result);
};
