// @payloadcms/db-mongodb/dist/findGlobal.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "findGlobal", {
    enumerable: true,
    get: function() {
        return findGlobal;
    }
});
const _database = require("payload/database");
const _sanitizeInternalFields = /*#__PURE__*/ _interop_require_default(require("./utilities/sanitizeInternalFields"));
const _withSession = require("./withSession");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const findGlobal = async function findGlobal({ slug, locale, req = {}, where }) {
    const Model = this.globals;
    const options = {
        ...await (0, _withSession.withSession)(this, req),
        lean: true
    };
    const query = await Model.buildQuery({
        globalSlug: slug,
        locale,
        payload: this.payload,
        where: (0, _database.combineQueries)({
            globalType: {
                equals: slug
            }
        }, where)
    });
    let doc = await Model.findOne(query, {}, options);
    if (!doc) {
        return null;
    }
    if (this.jsonParse) {
        doc = JSON.parse(JSON.stringify(doc));
    }
    if (doc._id) {
        doc.id = JSON.parse(JSON.stringify(doc._id));
        delete doc._id;
    }
    doc = (0, _sanitizeInternalFields.default)(doc);
    return doc;
};
