// @payloadcms/db-mongodb/dist/migrateFresh.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "migrateFresh", {
    enumerable: true,
    get: function() {
        return migrateFresh;
    }
});
const _database = require("payload/database");
const _commitTransaction = require("payload/dist/utilities/commitTransaction");
const _initTransaction = require("payload/dist/utilities/initTransaction");
const _killTransaction = require("payload/dist/utilities/killTransaction");
const _prompts = /*#__PURE__*/ _interop_require_default(require("prompts"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
async function migrateFresh({ forceAcceptWarning = false }) {
    const { payload } = this;
    if (!forceAcceptWarning) {
        const { confirm: acceptWarning } = await (0, _prompts.default)({
            name: 'confirm',
            type: 'confirm',
            initial: false,
            message: `WARNING: This will drop your database and run all migrations. Are you sure you want to proceed?`
        }, {
            onCancel: ()=>{
                process.exit(0);
            }
        });
        if (!acceptWarning) {
            process.exit(0);
        }
    }
    payload.logger.info({
        msg: `Dropping database.`
    });
    await this.connection.dropDatabase();
    const migrationFiles = await (0, _database.readMigrationFiles)({
        payload
    });
    payload.logger.debug({
        msg: `Found ${migrationFiles.length} migration files.`
    });
    const req = {
        payload
    };
    // Run all migrate up
    for (const migration of migrationFiles){
        payload.logger.info({
            msg: `Migrating: ${migration.name}`
        });
        try {
            const start = Date.now();
            await (0, _initTransaction.initTransaction)(req);
            await migration.up({
                payload,
                req
            });
            await payload.create({
                collection: 'payload-migrations',
                data: {
                    name: migration.name,
                    batch: 1
                },
                req
            });
            await (0, _commitTransaction.commitTransaction)(req);
            payload.logger.info({
                msg: `Migrated:  ${migration.name} (${Date.now() - start}ms)`
            });
        } catch (err) {
            await (0, _killTransaction.killTransaction)(req);
            payload.logger.error({
                err,
                msg: `Error running migration ${migration.name}. Rolling back.`
            });
            throw err;
        }
    }
}
