// @payloadcms/db-mongodb/dist/updateGlobalVersion.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "updateGlobalVersion", {
    enumerable: true,
    get: function() {
        return updateGlobalVersion;
    }
});
const _sanitizeInternalFields = /*#__PURE__*/ _interop_require_default(require("./utilities/sanitizeInternalFields"));
const _withSession = require("./withSession");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
async function updateGlobalVersion({ id, global, locale, req = {}, versionData, where }) {
    const VersionModel = this.versions[global];
    const whereToUse = where || {
        id: {
            equals: id
        }
    };
    const options = {
        ...await (0, _withSession.withSession)(this, req),
        lean: true,
        new: true
    };
    const query = await VersionModel.buildQuery({
        locale,
        payload: this.payload,
        where: whereToUse
    });
    let doc = await VersionModel.findOneAndUpdate(query, versionData, options);
    doc = this.jsonParse ? JSON.parse(JSON.stringify(doc)) : doc;
    const verificationToken = doc._verificationToken;
    if (verificationToken) {
        doc._verificationToken = verificationToken;
    }
    return (0, _sanitizeInternalFields.default)(doc);
}
