// @payloadcms/plugin-cloud/dist/utilities/getIncomingFiles.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "getIncomingFiles", {
    enumerable: true,
    get: function() {
        return getIncomingFiles;
    }
});
function getIncomingFiles({ data, req }) {
    const file = req.files?.file;
    let files = [];
    if (file && data.filename && data.mimeType) {
        const mainFile = {
            buffer: file.data,
            filename: data.filename,
            filesize: file.size,
            mimeType: data.mimeType,
            tempFilePath: file.tempFilePath
        };
        files = [
            mainFile
        ];
        if (data?.sizes) {
            Object.entries(data.sizes).forEach(([key, resizedFileData])=>{
                if (req.payloadUploadSizes?.[key] && data.mimeType) {
                    files = files.concat([
                        {
                            buffer: req.payloadUploadSizes[key],
                            filename: `${resizedFileData.filename}`,
                            filesize: req.payloadUploadSizes[key].length,
                            mimeType: data.mimeType
                        }
                    ]);
                }
            });
        }
    }
    return files;
}
